### PR TITLE
Add type hints to `synapse/storage/databases/main`

### PR DESCRIPTION
**Pull request by dklimpel**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: synapse/handlers/presence.py [hunk 1]
```diff
@@ -224,7 +225,9 @@ async def current_state_for_users(
                 states.update(new)
                 self.user_to_current_state.update(new)
 
-        return states
+        # mypy does not realize that is Dict[str, UserPresenceState]]
+        # and not Dict[str, Optional[UserPresenceState]]]
+        return cast(Dict[str, UserPresenceState], states)
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:230]: I'm not surprised, given the code as it stands is convoluted. TBH I'd be tempted to rewrite it as something like:

```python
states = {}
for user_id in user_ids:
    state = self.user_to_current_state.get(user_id, None)
    if state:
        states[user_id] = state
    else:
        ...
```

File: synapse/storage/databases/main/presence.py [hunk 2]
```diff
@@ -54,19 +62,21 @@ def __init__(
...
 
+        self._presence_id_gen: AbstractStreamIdGenerator
+
         if isinstance(database.engine, PostgresEngine):
             self._presence_id_gen = MultiWriterIdGenerator(
                 db_conn=db_conn,
                 db=database,
                 stream_name="presence_stream",
-                instance_name=self._instance_name,
+                instance_name=self._instance_name,  # type: ignore[attr-defined]
```
Comment (erikjohnston) [inline synapse/storage/databases/main/presence.py:79]: I'd be sorely tempted to do a `self._instance_name = hs.get_instance_name()` at the top, even though it technically is redefines it

File: synapse/storage/databases/main/user_directory.py [hunk 3]
```diff
@@ -234,10 +234,10 @@ def _get_next_batch(
         processed_event_count = 0
 
         for room_id, event_count in rooms_to_work_on:
-            is_in_room = await self.is_host_joined(room_id, self.server_name)
+            is_in_room = await self.is_host_joined(room_id, self.server_name)  # type: ignore[attr-defined]
```
Comment (erikjohnston) [inline synapse/storage/databases/main/user_directory.py:237]: Hmm, I'm quite uneasy about how many `# type: ignore[attr-defined]` there are here. If we're not careful we'll end up losing a lot of useful typing info. I'm not sure what the best way forward here is though, will canvas for opinions

Comment (erikjohnston) [reply synapse/storage/databases/main/user_directory.py:237]: Consensus was this was better than nothing

File: synapse/storage/databases/main/presence.py (hunk 2 above)
Comment (dklimpel) [reply synapse/storage/databases/main/presence.py]: Hopefully it's better now. The database query for a single user is not implemented `_get_presence_for_user`. My thoughts were that calling `get_presence_for_users` for every single user becomes a lot of overhead.
The loops are getting smaller now and perhaps increase the performance.

File: synapse/handlers/presence.py (hunk 1 above)
Comment (dklimpel) [reply synapse/handlers/presence.py]: Hopefully it's better now. The database query for a single user is not implemented `_get_presence_for_user`. My thoughts were that calling `get_presence_for_users` for every single user becomes a lot of overhead.
The loops are getting smaller now and perhaps increase the performance.

File: synapse/storage/databases/main/user_directory.py (hunk 3 above)
Comment (dklimpel) [reply synapse/storage/databases/main/user_directory.py:237]: I don't like it that much, but better to "ignore" the few lines and check the rest of the file.

File: synapse/handlers/presence.py [hunk 4]
```diff
@@ -204,25 +204,27 @@ async def current_state_for_users(
...
-            missing = [user_id for user_id, state in states.items() if not state]
-            if missing:
-                new = {
-                    user_id: UserPresenceState.default(user_id) for user_id in missing
-                }
-                states.update(new)
-                self.user_to_current_state.update(new)
+            for user_id in missing:
+                # if user has no state in database, create the state
+                if not res.get(user_id, None):
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:224]: Mildly more pythonic, and we don't have to worry about if a `UserPresenceState` can be falsey.

```suggestion
                if user_id not in res:
```

File: synapse/handlers/presence.py [hunk 5]
```diff
@@ -204,25 +204,27 @@ async def current_state_for_users(
...
+                    new = {user_id: UserPresenceState.default(user_id)}
+                    states.update(new)
+                    self.user_to_current_state.update(new)
```
Comment (erikjohnston) [inline synapse/handlers/presence.py:227]:
```suggestion
                    new_state = UserPresenceState.default(user_id)
                    states[user_id] = new_state
                    self.user_to_current_state[user_id] = new_state
```

Not much point making a new `dict` really

File: synapse/handlers/presence.py [hunk 6]
```diff
@@ -204,25 +204,27 @@ async def current_state_for_users(
...
-            missing = [user_id for user_id, state in states.items() if not state]
-            if missing:
-                new = {
-                    user_id: UserPresenceState.default(user_id) for user_id in missing
-                }
-                states.update(new)
-                self.user_to_current_state.update(new)
+            for user_id in missing:
+                # if user has no state in database, create the state
+                if user_id not in res:
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/presence.py]: Oh, bah. Sorry. Looks like `get_presence_for_users` returns `None` for missing users.

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**